README.md
# coleslaw

[![Build Status](https://travis-ci.org/kingcons/coleslaw.svg?branch=master)](https://travis-ci.org/kingcons/coleslaw)
[![Quicklisp](http://quickdocs.org/badge/coleslaw.svg)](http://quickdocs.org/coleslaw/)

<img src="https://raw.github.com/redline6561/coleslaw/master/themes/hyde/css/logo_medium.jpg" alt="coleslaw logo" align="right"/>

> [Czeslaw Milosz](http://blog.redlinernotes.com/tag/milosz.html) was the writer-in-residence at UNC c. 1992.
> I used to see him all the time at the Hardback Cafe, always sitting at a two-top
> drinking coffee, reading, writing, eating chips and salsa. I remember a gentleness
> behind the enormous bushy eyebrows and that we called him Coleslaw. - anon

Coleslaw is Flexible Lisp Blogware similar to [Frog](https://github.com/greghendershott/frog), [Jekyll](http://jekyllrb.com/), or [Hakyll](http://jaspervdj.be/hakyll/).

Have questions? Come talk to us on IRC in **#coleslaw** on Freenode!

## Features

* Git for storage
* RSS and Atom feeds
* Markdown Support with Code Highlighting provided by [colorize](http://www.cliki.net/colorize)
  * Currently supports: Common Lisp, Emacs Lisp, Scheme, C, C++, Java, Python, Erlang, Haskell, Obj-C, Diff.

* A [Plugin API](http://github.com/redline6561/coleslaw/blob/master/docs/plugin-api.md) and [**plugins**](http://github.com/redline6561/coleslaw/blob/master/docs/plugin-use.md) for...
  * Static Pages
  * Sitemap generation
  * Control of URL generation
  * Backup with symlinks, hardlinks, and/or git.
  * Incremental builds
  * Analytics via Google or [Piwik](http://www.piwik.org)
  * Comments via [Disqus](http://disqus.com/) or [isso](http://posativ.org/isso)
  * Hosting via [Github Pages](https://pages.github.com/) or [Amazon S3](http://aws.amazon.com/s3/)
  * Embedding [gfycats](http://gfycat.com/)
  * [Tweeting](http://twitter.com/) about new posts
  * Using LaTeX via [Mathjax](http://mathjax.org/)
  * Writing posts in ReStructured Text
  * Importing posts from [Wordpress](http://wordpress.org/)
  * Code Highlighting via [Pygments](http://pygments.org/) instead of [colorize](http://www.cliki.net/colorize)

## Example Sites

See the [wiki](https://github.com/redline6561/coleslaw/wiki/Blogroll) for a list of coleslaw-powered blogs.

## Hacking

A core goal of *coleslaw* is to be both pleasant to read and easy to
hack on and extend. If you want to understand the internals and bend
*coleslaw* to do new and interesting things, I strongly encourage you
to read the [Hacker's Guide to Coleslaw][hackers]. You'll find some
current **TODO** items towards the bottom.

[hackers]: https://github.com/redline6561/coleslaw/blob/master/docs/hacking.md

## Installation

Coleslaw should run on any conforming Common Lisp implementation but
testing is primarily done on [SBCL](http://www.sbcl.org/) and
[CCL](http://ccl.clozure.com/).

Coleslaw can either be run **manually** on a local machine or
triggered **automatically** on git push to a server.  If you want a
server install, run these commands on your server _after_ setting up a
[git bare repo](http://git-scm.com/book/en/Git-on-the-Server-Setting-Up-the-Server).
Otherwise, run the commands on your local machine.

1. Install a Common Lisp implementation (we recommend SBCL) and
   [Quicklisp](http://quicklisp.org/).
2. Place a config file for coleslaw in your `$HOME` directory. If you
   want to run multiple blogs with coleslaw, you can keep each blog's
   config file in that blog's repo.  Feel free to copy and edit the
   [example config][ex_config] or consult the [config docs][conf_docs]
   to create one from scratch.
3. This step depends on whether you're setting up a local or server install.
   * Server Install: Copy and `chmod +x` the
     [example post-receive hook][post_hook] to your blog's bare repo.
   * Local Install:  Just run the following commands in the
     REPL whenever you're ready to regenerate your blog:
     ```
     (ql:quickload :coleslaw)
     (coleslaw:main "/path/to/my/blog/")
     ```
4. Optionally, point the web server of your liking at your config-specified
   `:deploy-dir`. Or "deploy-dir/.curr" if the `versioned` plugin is enabled.

Now just write posts, git commit and build by hand or by push.

[ex_config]: https://github.com/redline6561/coleslaw/blob/master/examples/example.coleslawrc
[conf_docs]: https://github.com/redline6561/coleslaw/blob/master/docs/config.md
[post_hook]: https://github.com/redline6561/coleslaw/blob/master/examples/example.post-receive

## The Content Format

Coleslaw expects content to have a file extension matching the class
of the content. (I.e. `.post` for blog posts, `.page` for static
pages, etc.)

There should also be a metadata header on all files
starting and ending with the config-specified `:separator`, ";;;;;" by
default. Example:

```
;;;;;
title: foo
tags: bar, baz
date: yyyy-mm-dd hh:mm:ss
format: html (for raw html) or md (for markdown)
excerpt: Can also be extracted from content (see :excerpt-sep config param)
url: playing-with-coleslaw.html
;;;;;
your post
```

Posts require the `title:` and `format:` fields.
Pages require the `title:` and `url:` fields.

To omit a field, simply do not have the line present, empty lines and
fields (e.g. "tags:" followed by whitespace) will be ignored.

## Theming

Two themes are provided: hyde, the default, and readable (based on
[bootswatch readable](http://bootswatch.com/readable/)).

A guide to creating themes for coleslaw lives
[here](https://github.com/redline6561/coleslaw/blob/master/docs/themes.md).


changelog.md
# Changelog

## 0.2.5
- Pulled code from [Roman Hargrave](https://github.com/RomanHargrave "Thank you for your contribution.")'s branch.
- Made the dispatcher external.

## 0.2.4
- checkdoc, package-lint, flycheck all done for melpa submission.

## 0.2.3
- Dropped `M-;` keybinding, since coleslaw does what it needs to when opening a
  new file.

## 0.2.0
- Test framework in tests.el.
- Skeleton now does error checking on user input, and asks before inserting
  superfluous headers.
- Date auto-inserts the current date in ISO8601 (2000-01-01 style) format.

## 0.1.3 2019-06-16
- One-key interface (bound to `M-;`) now is
  `coleslaw-insert-header-or-dispatch`. Should only be needed when changing a
  files mode in the `format:`, to redispatch the modes.
- Easier control over mode dispatch using `coleslaw-default-format-modes`.
- Changed the lighted to `CSLAW`.
- Dispatchment on old files based on the `format:` field.
- Use different separators with `coleslaw-header-separator`.

## 0.0.1
- Skeleton on open, with mode dispatch (but no mode dispatch on nonfresh files).

# Contributors

## Author/Maintainer
Spenser Truex
## Pulled code
Roman Hargrave


tests.el
;;; tests.el --- Tests for coleslaw-mode -*- lexical-binding: t; -*-
;; Copyright (C) 2019 Spenser Truex
;; Author: Spenser Truex <web@spensertruex.com>
;; Created: 2019-06-16 
;; Package-Requires: ((emacs "24"))
;; URL: https://github.com/equwal/coleslaw/
;; Homepage: https://spensertruex.com/coleslaw-mode
;; This file is not part of GNU Emacs.
;; This file is free software.
;; License:
;; Licensed with the GNU GPL v3 see:
;; <https://www.gnu.org/licenses/>.
;;; Commentary:
;; Tests to run for coleslaw-mode. Main code in coleslaw.el

(require 'coleslaw)
(defvar coleslaw--tests nil "Test functions for coleslaw. Return T or failed.")

(defmacro coleslaw-add-test (name &rest body)
  `(add-to-list 'coleslaw--tests
                '(,name (lambda () ,@body))
                t
                (lambda (x y) (and (listp x)
                                   (listp y)
                                   (eql (car x) (car y))))))

(coleslaw-add-test coleslaw--true t)
(coleslaw-add-test coleslaw--false nil)


;;;###autoload
(defun coleslaw-tests ()
  (let ((res (cl-loop for test in coleslaw--tests
                      collect (let ((success (funcall (cadr test))))
                                (if (not (eql success t))
                                    (format "Test %s failed with %S."
                                            (symbol-name (car test))
                                            success)
                                  success)))))
    (remove-if (lambda (x) (eql t x)) res)
    ;; (if (car res) res t)
    ))
(provide 'coleslaw-tests)
;;; tests.el ends here


coleslaw.el
;;; coleslaw.el --- Coleslaw static content files. -*- lexical-binding: t; -*-
;; Copyright (C) 2019 Spenser Truex
;; Author: Spenser Truex <web@spensertruex.com>
;; Created: 2019-06-16
;; Version: 0.2.5
;; Package-Requires: ((emacs "24.4"))
;; Keywords: lisp wp files convenience
;; URL: https://github.com/equwal/coleslaw/
;; Homepage: https://spensertruex.com/coleslaw-mode
;; This file is not part of GNU Emacs, but you want to use  GNU Emacs to run it.
;; This file is very free software.
;; License:
;; Licensed with the GNU GPL v3 see:
;; <https://www.gnu.org/licenses/>.

;;; Commentary:

;; Please add (coleslaw-setup) to your init file for the author's mode
;; selections. Bind `coleslaw-dispatch' to a key to make coleslaw discover the
;; mode.

;; For the coleslaw static content generator, a minor mode which inserts the
;; header, selects the major mode, and generally makes writing static content
;; easier.

;; Consider also installing the coleslaw-snippet package to generate your snippets.
;;; Code:

(require 'cl-lib)

(defvar coleslaw-mode-hook nil "Coleslaw-mode, for editing static web content.")
(defvar coleslaw-separator ";;;;;" "Character that separates keys and values in the coleslaw header")

(defvar coleslaw-modes nil
  (concat
   "Modes based on the regex (special characters quoted)"
   (regexp-quote coleslaw-separator)
   "
  format: FORMAT
" (regexp-quote coleslaw-separator) "
  headers in the coleslaw file. A simple default choice is:
  (setq coleslaw-modes
        '((\"md\" . (markdown-mode))
          (\"cl-who\" . (lisp-mode))
          (\"html\" . (html-mode))
          (\"rst\" . (rst-mode))))
  in your init file."))

(add-hook 'coleslaw-mode-hook 'coleslaw-dispatch)

;;;###autoload
(defun coleslaw-setup ()
  "Setup your coleslaw like the author suggests (conservative edits only).
strongly recommended!  Enable auto insertion for .page and .post
files, enable such basic editing modes as the mode function
`markdown-mode', the mode function `lisp-mode', the mode function
`html-mode', or the mode function `rst-mode' based on the format
header field.  Conservative additions only."
  (setq coleslaw-modes
        '(("md" . (markdown-mode))
          ("cl-who" . (lisp-mode))
          ("html" . (html-mode))
          ("rst" . (rst-mode)))))

(defun coleslaw-dispatch-format (format)
  "Select the mode for a file of type FORMAT."
  (mapc (lambda (mode) (funcall mode))
        (cdr (assoc format coleslaw-modes #'string-equal))))

(defun coleslaw-dispatch ()
  "Set modes based on this buffer's 'format: (md, cl-who, etc.)' metadata line."
  (interactive)
  (when (coleslaw--header-detected)
    (coleslaw-dispatch-format (coleslaw--header-field "format"))))

(defun coleslaw--re-search-whole (regex &optional bound noerror count)
  "Search forward and backwards from the point in the buffer for REGEX.
BOUND, NOERROR, and COUNT are arguments to `re-search-forward'
and `re-search-backward'."
  (let ((args (list regex bound noerror count)))
    (if (apply #'re-search-forward args)
        (match-string 1)
      (when (apply #'re-search-backward args)
        (match-string 1)))))

(defun coleslaw--header-detected ()
  "Detect if a header is already in the file."
  ;; pointer in fields, under fields, or in separator
  (or (re-search-forward (regexp-quote coleslaw-separator)
                         nil t 1)
      (re-search-backward (regexp-quote coleslaw-separator)
                          nil t)))

(defun coleslaw--header-field (field)
  "Search the current bufffer for the header FIELD.
Don't include the colon in the FIELD string (e.g. \"format\")."
  (when (coleslaw--header-detected)
    (coleslaw--re-search-whole (concat field
                                       ":"
                                       "[\t ]*\\(?1:.*\\)\n")
                               nil t)))

;;;###autoload
(define-minor-mode coleslaw-mode "Edit coleslaw static content gloriously."
  :lighter " Coleslaw"
  (dolist (type '("\\.page\\'" "\\.post\\'"))
    (add-to-list 'auto-mode-alist (cons type 'coleslaw-mode)))
  (coleslaw-dispatch))

(provide 'coleslaw)
;;; coleslaw.el ends here


src/config.lisp
(in-package :coleslaw)

(defclass blog ()
  ((author          :initarg :author         :reader author)
   (charset         :initarg :charset        :reader charset)
   (deploy-dir      :initarg :deploy-dir     :reader deploy-dir)
   (domain          :initarg :domain         :reader domain)
   (excerpt-sep     :initarg :excerpt-sep    :reader excerpt-sep)
   (feeds           :initarg :feeds          :reader feeds)
   (name-fn         :initarg :name-fn        :reader name-fn)
   (lang            :initarg :lang           :reader lang)
   (license         :initarg :license        :reader license)
   (page-ext        :initarg :page-ext       :reader page-ext-intolerant)
   (plugins         :initarg :plugins        :reader plugins)
   (repo            :initarg :repo           :accessor repo-dir)
   (routing         :initarg :routing        :reader routing)
   (separator       :initarg :separator      :reader separator)
   (sitenav         :initarg :sitenav        :reader sitenav)
   (staging-dir     :initarg :staging-dir    :reader staging-dir)
   (theme           :initarg :theme          :reader theme)
   (title           :initarg :title          :reader title)
   (index-ext       :initarg :index-ext      :reader index-ext)
   (conf-sitemap    :initarg :conf-sitemap   :reader conf-sitemap)
   (rsync-passfile  :initarg :rsync-passfile :reader rsync-passfile)
   (which-sshpass   :initarg :which-sshpass  :reader which-sshpass))
  
  (:default-initargs
   :feeds        nil
   :license      nil
   :conf-sitemap nil
   :plugins      nil
   :sitenav      nil
   :rsync-passfile nil
   :which-sshpass nil
   :excerpt-sep  "<!--more-->"
   :name-fn      'identity
   :charset      "UTF-8"
   :lang         "en"
   :page-ext     #1="html"
   :separator    ";;;;;"
   :staging-dir  "/tmp/coleslaw"
   :index-ext    #1#))

(defun page-ext (config)
  "Get page extension, and allow for an extensionless system."
  (let ((ext (page-ext-intolerant config)))
    (if (string= ext "")
        ""
        (concatenate 'string "." ext))))
(defun dir-slot-reader (config name)
  "Take CONFIG and NAME, and return a directory pathname for the matching SLOT."
  (ensure-directory-pathname (slot-value config name)))

(defmethod deploy-dir  ((config blog)) (dir-slot-reader config 'deploy-dir))
(defmethod repo-dir    ((config blog)) (dir-slot-reader config 'repo))
(defmethod staging-dir ((config blog)) (dir-slot-reader config 'staging-dir))

(defparameter *config* nil
  "A variable to store the blog configuration and plugin settings.")

(define-condition plugin-conf-error ()
  ((plugin  :initarg :plugin :reader plugin)
   (message :initarg :message :reader message))
  (:report (lambda (condition stream)
             (format stream "~A: ~A" (plugin condition) (message condition))))
  (:documentation "Condition to signal when the plugin is misconfigured."))

(defun enable-plugin (name args)
  "Given a plugin, NAME, compile+load it and call its ENABLE function with ARGS."
  (flet ((plugin-path (sym)
           (if (probe-file (repo-path "plugins/~(~A~).lisp" sym))
               (repo-path "plugins/~(~A~)" sym)
               (app-path "plugins/~(~A~)" sym)))
         (plugin-package (sym)
           (format nil "~:@(coleslaw-~A~)" sym)))
    (let ((file (plugin-path name)))
      (multiple-value-bind (output-file error)
          (ignore-errors (compile-file file :verbose nil :print nil))
        (when error
          (warn "Error while compiling plugin ~A: ~A.~%" name error))
        (load (or output-file file) :verbose t)))
    (let ((package (find-package (plugin-package name))))
      (apply (find-symbol "ENABLE" package) args))))

(defun load-plugins (plugins)
  "Compile and load the listed PLUGINS. It is expected that matching *.lisp files
are in the plugins folder in coleslaw's source directory."
  (setf *injections* nil)
  (dolist (plugin plugins)
    (destructuring-bind (name &rest args) plugin
      (enable-plugin name args))))

(defun discover-config-path (repo-path)
  "Check the supplied REPO-PATH for a .coleslawrc and if one
doesn't exist, use the .coleslawrc in the home directory."
  (let ((repo-config (rel-path repo-path ".coleslawrc")))
    (if (file-exists-p repo-config)
        repo-config
        (rel-path (user-homedir-pathname) ".coleslawrc"))))

(defun load-config (&optional (repo-dir ""))
  "Find and load the coleslaw configuration from .coleslawrc. REPO-DIR will be
preferred over the home directory if provided."
  (with-open-file (in (discover-config-path repo-dir) :external-format :utf-8)
    (let ((config-form (read in)))
      (setf *config* (construct 'blog config-form)
            (repo-dir *config*) repo-dir)))
  (load-plugins (plugins *config*)))


coleslaw-mode.el
;;; coleslaw-mode.el --- Coleslaw static content files. -*- lexical-binding: t; -*-
;; Copyright (C) 2018 Spenser Truex
;; Author: Spenser Truex <web@spensertruex.com>
;; Created: 2019-06-16
;; Version: 0.2.4
;; Package-Requires: ((emacs "24"))
;; Keywords: lisp wp files convenience
;; URL: https://github.com/equwal/coleslaw/
;; Homepage: https://spensertruex.com/coleslaw-mode
;; This file is not part of GNU Emacs, but you want to use  GNU Emacs to run it.
;; This file is very free software.
;; License:
;; Licensed with the GNU GPL v3 see:
;; <https://www.gnu.org/licenses/>.

;;; Commentary:
;; Please add (coleslaw-setup) to your init file for the author's mode
;; selections.

;; For the coleslaw static content generator, a minor mode which inserts the
;; header, selects the major mode, and generally makes writing static content
;; easier.

;;; Code:

(defvar coleslaw-mode-hook nil "Coleslaw-mode, for editing static web content.")

(defvar coleslaw-formats (list "md" "cl-who" "rst" "html" "org")
  "The format header values that coleslaw will allow to be auto-inserted.")

(defvar coleslaw-separator ";;;;;"
  "The string used between the coleslaw headers as in the example:
;;;;;
title: Example
format: cl-who
date: 2019-06-15
;;;;;
Where the separator is \";;;;;\".")

(defvar coleslaw-auto-insert (when (boundp 'coleslaw-auto-insert)
                               coleslaw-auto-insert)
  "Predicate to insert the skeleton on opening a new Coleslaw file type.
unless the function `coleslaw-setup' is ran, when it is set to T.")

(defvar coleslaw-modes nil
  (concatenate 'string
               "Modes based on the regex (special characters quoted)"
               (regexp-quote coleslaw-separator)
               "
  format: FORMAT
" (regexp-quote coleslaw-separator) "
  headers in the coleslaw file. A simple default choice is:
  (setq coleslaw-modes
        '((\"md\" . (markdown-mode))
          (\"cl-who\" . (lisp-mode))
          (\"html\" . (html-mode))
          (\"rst\" . (rst-mode))))
  in your init file."))

(defun coleslaw--valid-format (str)
  "Determine if the STR is permissible for a format: header in Coleslaw."
  (when (stringp str)
    (some (lambda (x) (string-equal x str)) coleslaw-formats)))

(defun coleslaw-setup ()
  "Setup your coleslaw like the author suggests (conservative edits only).
strongly recommended!  Enable auto insertion for .page and .post
files, enable such basic editing modes as the mode function
`markdown-mode', the mode function `lisp-mode', the mode function
`html-mode', or the mode function `rst-mode' based on the format
header field.  Conservative additions only."
  (when (require 'autoinsert nil 'installed)
    (setq coleslaw-auto-insert t)
    (dolist (type '(".page" ".post"))
      (add-to-list 'auto-insert-alist (cons type 'coleslaw-insert-header))))
  (dolist (type '("\\.page\\'" "\\.post\\'"))
    (add-to-list 'auto-mode-alist (cons type 'coleslaw-mode)))
  (add-hook 'coleslaw-mode-hook 'coleslaw--dispatch)
  (when (not coleslaw-modes)
    (setq coleslaw-modes
          '(("md" . (markdown-mode))
            ("cl-who" . (lisp-mode))
            ("html" . (html-mode))
            ("rst" . (rst-mode))))))

(defun coleslaw--bufftype (type)
  "Determine if the file type of the current buffer is TYPE."
  (string-equal type (cl-subseq buffer-file-name (- (length buffer-file-name) 5))))

(defun coleslaw--mode-spawn (format)
  "Select the mode for a file of type FORMAT."
  (mapc (lambda (mode) (funcall mode))
        (cdr (assoc format coleslaw-modes #'string-equal))))

(defun coleslaw--dispatch ()
  "Set modes based on this buffer's 'format: (md, cl-who, etc.)' metadata line."
  (when (coleslaw--header-detected)
    (coleslaw--mode-spawn (coleslaw--header-field "format"))))

(defun coleslaw--insist-format (first-prompt &optional second-prompt)
  "Insist that the format inserted be a valid coleslaw format.
Add formats to the `coleslaw-formats' list for new features or
nonstandard formats.  FIRST-PROMPT and SECOND-PROMPT are used to
prompt the user at the minibuffer.  The FIRST-PROMPT is for the
first, and SECOND-PROMPT is used in subsequent requests.  If
FIRST-PROMPT is NIL, the second-prompt is only used."
  (let ((res (read-from-minibuffer (if first-prompt
                                       first-prompt
                                     second-prompt))))
    (if (not (coleslaw--valid-format res))
        (coleslaw--insist-format nil second-prompt))))
;;;###autoload
(defun coleslaw-insert-header  ()
  "Insert the skeleton for as specified by default for a coleslaw file type."
  (skeleton-insert '(nil str
                         "\ntitle: "
                         (skeleton-read "title: ")
                         "\nformat: "
                         (coleslaw--insist-format "format: "
                                                  "Bad format, try another format: ")
                         (if (coleslaw--bufftype ".page")
                             (concat "\nurl: " (skeleton-read "url: "))
                           "")
                         (if (coleslaw--bufftype ".post")
                             (if (y-or-n-p "Insert excerpt? ")
                                 (concat "\nexcerpt: "
                                         (skeleton-read "excerpt: "))
                               "")
                           "")
                         "\ndate: "
                         (format-time-string "%Y-%m-%d" (current-time))
			 "\n" str)
		   0 (regexp-quote coleslaw-separator))
  (coleslaw--dispatch))

(defun coleslaw--re-search-whole (regex &optional bound noerror count)
  "Search forward and backwards from the point in the buffer for REGEX.
BOUND, NOERROR, and COUNT are arguments to `re-search-forward'
and `re-search-backward'."
  (let ((args (list regex bound noerror count)))
    (if (apply #'re-search-forward args)
        (match-string 1)
      (when (apply #'re-search-backward args)
        (match-string 1)))))

(defun coleslaw--header-detected ()
  "Detect if a header is already in the file."
  ;; pointer in fields, under fields, or in separator
  (or (re-search-forward (regexp-quote coleslaw-separator)
                         nil t 1)
      (re-search-backward (regexp-quote coleslaw-separator)
                          nil t)))

(defun coleslaw--header-field (field)
  "Search the current bufffer for the header FIELD.
Don't include the colon in the FIELD string (e.g. \"format\")."
  (when (coleslaw--header-detected)
    (coleslaw--re-search-whole (concatenate 'string
                                            field
                                            ":"
                                            "[\t ]*\\(?1:.*\\)\n")
                               nil t)))

;;;###autoload
(define-minor-mode coleslaw-mode "Edit coleslaw static content gloriously."
  :lighter " Coleslaw"
  (when (require 'autoload) (auto-insert))
  (coleslaw--dispatch))

(provide 'coleslaw-mode)
;;; coleslaw-mode.el ends here
